src/objc/ModalWebViewController.mm

```diff
@@ -115,6 +99,36 @@ - (void)viewDidDisappear:(BOOL)animated {
   }
 }
 
+- (void)observeValueForKeyPath:(NSString *)keyPath
+                      ofObject:(id)object
+                        change:(NSDictionary<NSKeyValueChangeKey,id> *)change
+                       context:(void *)context {
+  if ([keyPath isEqualToString:@"URL"] && object == self.webView) {
+    NSURL *newURL = change[NSKeyValueChangeNewKey];
+    if (newURL && [newURL isKindOfClass:[NSURL class]]) {
+      NSDictionary *event = @{
+        @"event": @"location_changed",
+        @"url": newURL.absoluteString ?: @"",
+        @"id": [NSString stringWithFormat:@"%.0f", [[NSDate date] timeIntervalSince1970] * 1000]
+      };
+
+      NSError *error = nil;
+      NSData *jsonData = [NSJSONSerialization dataWithJSONObject:event options:0 error:&error];
+
+      if (jsonData) {
+        NSString *payload = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
+          
+        [self addToVisitedUrls:newURL.absoluteString];
```

Comment:
I guess this can be outside the if? which also has a null check on line 111? `@"url": newURL.absoluteString ?: @""` wouldn't it be better to check at the beginning of the function?